### DISCUSSION TITLE
WPF Control, How to set transparency?

### DISCUSSION THREAD

**Discussion by rafayahmed317** [microsoft/terminal#15989] [Q&A] [answered]
I know the work is going on for the WPF and UWP controls, but I was looking on how to set the transparency for the WPF terminal control. Currently, it seems that the theme allows only for a solid background color, so I modified the TerminalSetTheme Method to also do EnableTransparentBackground(true) , however still the terminal control loads using the Background solid color, I looked up the source code for the Terminal app to see how it was doing this, and aside for the EnableTransparentBackground setting (I dont know c++. so I certainly have missed something), I was unable to locate how it did this. That means that I could not find where the transparency setting should be made, does the renderer does this, or will I have to modify something else?

**Comment by zadjii-msft** [comment]
How exactly are you running the WPF control? It's not really _super_ well supported, certainly not as a standalone app right now. 

I'd suspect that even if you set the transparency of the control itself, it would just be transparent to the Background of the app window itself. IIRC it's pretty trivial to change that in WPF: https://stackoverflow.com/a/21461482/1481137

https://github.com/microsoft/terminal/blob/main/src/cascadia/WpfTerminalTestNetCore/MainWindow.xaml

**Reply by rafayahmed317** [reply]
Yeah I did that. It appears the background of the control is set through its theme, qnd the theme does not support transparency. I just wanted to know how is it handled in the uwp app.

**Reply by zadjii-msft** [reply]
I mean, you're definitely on the right track with `ControlCore::_setOpacity`.

In the UWP XAML app, we've had to really go through and make sure that absolutely every later was set to `Transparent`, so that really the only layer that had color was the `TermControl` itself. I honestly don't think in the WPF version we've done the same diligence yet. 

I reckon it's possible that the HWND for the WPF version might need some additional teaks too. There's a call to `CreateWindowExW` in `HwndTerminal::HwndTerminal`. That's not passing `WS_EX_NOREDIRECTIONBITMAP` like `IslandWindow::MakeWindow` does. Maybe that's it?

**Reply by rafayahmed317** [reply]
Thanks for the clues, I will try it and let you know.

**Reply by rafayahmed317** [reply]
I set the CreateWindowExW method to pass WS_EX_NOREDIRECTIONBITMAP but, it doesn't really change things, because we don't have anywhere to set the opacity, and in the HWNDHOST, the opacity has really no effect. Like in the TermControl, we have the _setOpacity function, what will it take for me to implement that? Sorry, if I am bothering you.

**Comment by rafayahmed317** [comment]
So I searched the internet and found [this](https://www.codeproject.com/Articles/5475/Quick-and-Dirty-Window-Transparency), and I was able to make the window transparent by making it layered and then changing its attributes, but there is a weird problem now, if I call this in the CreateTerminalFunction, it works, however from any other function after the terminal has been created such as from TerminalSetTheme, it does not seem to work, here is the code:
```
BOOL MakeWindowTransparent(HWND hWnd, unsigned char factor)
{
    /* First, see if we can get the API call we need. If we've tried
    * once, we don't need to try again. */
    if (!initialized)
    {
        HMODULE hDLL = LoadLibrary(L"user32");

        pSetLayeredWindowAttributes =
            (PSLWA)GetProcAddress(hDLL, "SetLayeredWindowAttributes");

        initialized = TRUE;
    }

    if (pSetLayeredWindowAttributes == NULL)
        return FALSE;

    /* Windows need to be layered to be made transparent. This is done
    * by modifying the extended style bits to contain WS_EX_LAYARED. */
    SetLastError(0);

    SetWindowLong(hWnd,
                  GWL_EXSTYLE,
                  GetWindowLong(hWnd, GWL_EXSTYLE) | WS_EX_LAYERED);

    if (GetLastError())
        return FALSE;

    /* Now, we need to set the 'layered window attributes'. This
    * is where the alpha values get set. */
    return pSetLayeredWindowAttributes(hWnd,
                                       RGB(255, 255, 255),
                                       factor,
                                       LWA_COLORKEY | LWA_ALPHA);
}
```
and here is the CreateTerminal function in which it works:
```
HRESULT _stdcall CreateTerminal(HWND parentHwnd, _Out_ void** hwnd, _Out_ void** terminal)
{
    auto _terminal = std::make_unique<HwndTerminal>(parentHwnd);
    RETURN_IF_FAILED(_terminal->Initialize());
    MakeWindowTransparent(_terminal->GetHwnd(), 50);
    *hwnd = _terminal->GetHwnd();
    *terminal = _terminal.release();

    return S_OK;
}
```
And here is the SetTransparency function, I created a new one, from which it does not work (It does not from any other function such as TerminalSetTheme) :
```
void _stdcall SetTransparency(void* terminal, unsigned char Transparency) {
    const auto publicTerminal = static_cast<HwndTerminal*>(terminal);
    if (!publicTerminal || !publicTerminal->_terminal)
    {
        return;
    }
    {
        auto lock = publicTerminal->_terminal->LockForWriting();
        MakeWindowTransparent(publicTerminal->GetHwnd(), Transparency);
        InvalidateRect(publicTerminal->GetHwnd(), NULL, TRUE);
        UpdateWindow(publicTerminal->GetHwnd());
        publicTerminal->_renderer->NotifyPaintFrame();
        publicTerminal->_renderer->TriggerRedrawAll(true, true);
    }
}
```
Please don't mind if I am making a noob error here, because cpp is not my language.

**Reply by rafayahmed317** [reply]
While debugging my issue, I also found out that maybe the renderer is not doing the same thing in the terminal app and in the wpf control, 
 dunno weather it makes a difference or not but in wpf the swapchain is swapchain hwnd and in the terminal app it is different.

**Reply by rafayahmed317** [reply]
@zadjii-msft Right now, I am able to set the transparency, but only when the window gets created and only by changing layered window attributes, and if I do that, then the foreground's opacity is changed as well, do you know any other method?

**Reply by zadjii-msft** [reply]
> but in wpf the swapchain is swapchain hwnd and in the terminal app it is different

That honestly might be something - I'm not sure I know enough about the intricacies of the difference between the HWND vs Composition swapchains to know if that's the root cause. 

> changing layered window attributes, and if I do that, then the foreground's opacity is changed as well

Yea, that sounds about right. That's the reason the `conhost` window behaves the same way (IIRC).

**Comment by zadjii-msft** [comment]
I had another thought here this morning - are we totally sure that the WPF stack is passing colors that have a transparent alpha channel to the renderer? IIRC `Terminal` used to have some dedicated code to ensure that the "default background" color always had `0` for the alpha channel. (for `COLORREF`s, I believe it's laid out as `0xAABBGGRR`)

I can't find that code anymore - we've gone through a bunch of refactors since that was written. But I dunno if WPF ever considered that.

**Reply by DHowett** [reply]
What WPF is and is not capable of is largely unrelated to what the Terminal component is capable of. We don't get any of our configuration through WPF, and are not subject to any limitations--if any exist--in its color representation.

We take in a uint32 in BGRx for each of the foreground/background/ANSI colors. We never plumbed through alpha because nobody requested it 😅.

**Comment by rafayahmed317** [comment]
WPF always passes RGB I think.

**Comment by rafayahmed317** [comment]
There was no option to pass the Alpha channel.

**Comment by DHowett** [comment]
A lot of this stuff is going to change with the new converged control platform over in https://github.com/microsoft/terminal/tree/dev/duhowett/win7-wpf-termcontrol. It might be better to check back once that merges!

In short: right now the WPF control uses the lower level components of Terminal like the renderer and the VT engine. Because of that, we've had to implement higher-level behaviors like selection and scrolling and color handling (to name a few) two times, and for the most part left the WPF control in stasis. Not intentionally, mind you! Just because it's easy to forget something that isn't built into the core product.

In shorter: we're combining the two Terminal controls into one library so they get the same features at the same time.

**Reply by rafayahmed317** [reply]
Thanks, I will keep an eye on it.

**Comment by rafayahmed317** [comment, marked as answer]
Finally, I was able to achieve better looking transparency by incorporating DirectComposition in the renderer code. But that is not available on Windows 7. But at least it will work fine on Win 8 and beyond.

**Reply by mitchcapper** [reply]
Can you share the code you were using for this?

**Reply by rafayahmed317** [reply]
@mitchcapper I don't have the code lying around, but I had to edit the rendering code to use Direct Composition and recompile that and rebuild the WPF related packages.

**Reply by mitchcapper** [reply]
I am guessing along the lines of https://github.com/riley-x/TransparentWindow

**Reply by rafayahmed317** [reply]
@mitchcapper Yes, I think I used this project as reference.

**Comment by slezisatoshiz** [comment]
The best!

**Reply by rafayahmed317** [reply]
The best?
